File: 0c-bason/assignment-1/cpp/src/flatten.cpp
#include "flatten.h"
#include "path_utils.h"
#include "ron64.h"

#include <map>
#include <stdexcept>

////////////////////////////////////////////////////////////////////////////////

namespace NBason {

////////////////////////////////////////////////////////////////////////////////

namespace {

void FlattenRecursive(
    const TBasonRecord& record,
    const std::string& currentPath,
    std::vector<TBasonRecord>& result)
{
    if (record.Type == EBasonType::Array || record.Type == EBasonType::Object) {
        // Container - recurse into children
        for (const auto& child : record.Children) {
            std::string childPath = currentPath.empty() 
                ? child.Key 
                : currentPath + "/" + child.Key;
            FlattenRecursive(child, childPath, result);
        }
    } else {
        // Leaf - add to result
        TBasonRecord flat(record.Type, currentPath, record.Value);
        result.push_back(std::move(flat));
    }
}

struct TTreeNode {
    EBasonType Type = EBasonType::Object;
    std::string Value;
    std::map<std::string, TTreeNode> Children;
    bool IsLeaf = false;
};

void BuildTree(const std::vector<TBasonRecord>& records, TTreeNode& root)
{
    for (const auto& record : records) {
        auto segments = SplitPath(record.Key);
        
        if (segments.empty()) {
            // Root-level leaf
            root.Type = record.Type;
            root.Value = record.Value;
            root.IsLeaf = true;
            continue;
        }

        TTreeNode* current = &root;
        
        for (size_t i = 0; i < segments.size(); ++i) {
            const auto& segment = segments[i];
            bool isLast = (i == segments.size() - 1);
            
            if (isLast) {
                // Leaf node
                auto& node = current->Children[segment];
                node.Type = record.Type;
                node.Value = record.Value;
                node.IsLeaf = true;
            } else {
                // Intermediate node - determine if it's array or object
                auto& node = current->Children[segment];
                if (node.Children.empty() && !node.IsLeaf) {
                    // Try to parse next segment as RON64 to detect arrays
                    if (i + 1 < segments.size()) {
                        try {
                            DecodeRon64(segments[i + 1]);
                            node.Type = EBasonType::Array;
                        } catch (...) {
                            node.Type = EBasonType::Object;
                        }
                    }
                }
                current = &node;
            }
        }
    }
}

void TreeToRecord(const TTreeNode& node, const std::string& key, TBasonRecord& record)
{
    record.Key = key;
    record.Type = node.Type;
    
    if (node.IsLeaf) {
        record.Value = node.Value;
    } else {
        // Container
        for (const auto& [childKey, childNode] : node.Children) {
            TBasonRecord child;
            TreeToRecord(childNode, childKey, child);
            record.Children.push_back(std::move(child));
        }
    }
}

} // namespace

////////////////////////////////////////////////////////////////////////////////

std::vector<TBasonRecord> FlattenBason(const TBasonRecord& root)
{
    std::vector<TBasonRecord> result;
    FlattenRecursive(root, root.Key, result);
    return result;
}

TBasonRecord UnflattenBason(const std::vector<TBasonRecord>& records)
{
    if (records.empty()) {
        return TBasonRecord(EBasonType::Object);
    }

    TTreeNode tree;
    BuildTree(records, tree);

    TBasonRecord result;
    TreeToRecord(tree, "", result);
    
    return result;
}

////////////////////////////////////////////////////////////////////////////////

} // namespace NBason

File: 0c-bason/assignment-1/go/pkg/bason/flatten.go
package bason

import "sort"

// FlattenBason converts a nested BASON record into a sequence of flat path-keyed leaf records.
func FlattenBason(root Record) []Record {
	var result []Record
	flattenRecursive(root, root.Key, &result)
	return result
}

func flattenRecursive(record Record, currentPath string, result *[]Record) {
	if record.Type == TypeArray || record.Type == TypeObject {
		for _, child := range record.Children {
			childPath := currentPath
			if childPath != "" {
				childPath += "/" + child.Key
			} else {
				childPath = child.Key
			}
			flattenRecursive(child, childPath, result)
		}
	} else {
		*result = append(*result, Record{
			Type:  record.Type,
			Key:   currentPath,
			Value: record.Value,
		})
	}
}

type treeNode struct {
	Type     BasonType
	Value    string
	Children map[string]*treeNode
	IsLeaf   bool
}

func buildTree(records []Record) *treeNode {
	root := &treeNode{Type: TypeObject, Children: make(map[string]*treeNode)}
	for _, record := range records {
		segments := SplitPath(record.Key)
		if len(segments) == 0 {
			root.Type = record.Type
			root.Value = record.Value
			root.IsLeaf = true
			continue
		}
		current := root
		for i, segment := range segments {
			isLast := i == len(segments)-1
			node, ok := current.Children[segment]
			if !ok {
				node = &treeNode{Children: make(map[string]*treeNode)}
				current.Children[segment] = node
			}
			if isLast {
				node.Type = record.Type
				node.Value = record.Value
				node.IsLeaf = true
			} else {
				if len(node.Children) == 0 && !node.IsLeaf && i+1 < len(segments) {
					if _, err := DecodeRon64(segments[i+1]); err == nil {
						node.Type = TypeArray
					} else {
						node.Type = TypeObject
					}
				}
				current = node
			}
		}
	}
	return root
}

func treeToRecord(node *treeNode, key string) Record {
	record := Record{Key: key, Type: node.Type}
	if node.IsLeaf {
		record.Value = node.Value
	} else {
		keys := make([]string, 0, len(node.Children))
		for k := range node.Children {
			keys = append(keys, k)
		}
		sort.Strings(keys)
		for _, childKey := range keys {
			record.Children = append(record.Children, treeToRecord(node.Children[childKey], childKey))
		}
	}
	return record
}

// UnflattenBason converts a sequence of flat path-keyed leaf records into a nested BASON record tree.
func UnflattenBason(records []Record) Record {
	if len(records) == 0 {
		return Record{Type: TypeObject}
	}
	tree := buildTree(records)
	return treeToRecord(tree, "")
}
